runtime/dancebox/tests/integration_test.rs

```diff
@@ -5527,3 +5527,311 @@ fn test_migration_services_collator_assignment_payment() {
+            (1003, empty_genesis_data(), vec![], u32::MAX, u32::MAX).into(),
+        ])
+        .build()
+        .execute_with(|| {
+            let tank_funds = 1 * UNIT;
+            let max_tip = 1 * UNIT;
+            let para_id = 1003u32;
+
+            // Send insufficient funds to tank for tip for 2 sessions
+            assert_ok!(ServicesPayment::purchase_credits(
+                origin_of(ALICE.into()),
+                para_id.into(),
+                tank_funds,
+            ));
+
+            // Set tip for para_id
+            assert_ok!(ServicesPayment::set_max_tip(
+                root_origin(),
+                para_id.into(),
+                max_tip,
+            ));
+
+            run_to_session(1);
+            assert_eq!(
+                CollatorAssignment::collator_container_chain().container_chains[&para_id.into()]
+                    .len(),
+                0
+            );
+        });
+}
+
+#[test]
+fn test_collator_assignment_tip_only_charge_willing_paras() {
+    ExtBuilder::default()
+        .with_balances(vec![
+            (AccountId::from(ALICE), 210_000 * UNIT),
+            (AccountId::from(BOB), 100_000 * UNIT),
+            (AccountId::from(CHARLIE), 100_000 * UNIT),
+            (AccountId::from(DAVE), 100_000 * UNIT),
+            (AccountId::from(EVE), 100_000 * UNIT),
+            (AccountId::from(FERDIE), 100_000 * UNIT),
+        ])
+        .with_collators(vec![
+            (AccountId::from(ALICE), 210 * UNIT),
+            (AccountId::from(BOB), 100 * UNIT),
+            (AccountId::from(CHARLIE), 100 * UNIT),
+            (AccountId::from(DAVE), 100 * UNIT),
+            (AccountId::from(EVE), 100 * UNIT),
+            (AccountId::from(FERDIE), 100 * UNIT),
+        ])
+        .with_para_ids(vec![
+            (1001, empty_genesis_data(), vec![], u32::MAX, u32::MAX).into(),
+            (1002, empty_genesis_data(), vec![], u32::MAX, u32::MAX).into(),
+            (1003, empty_genesis_data(), vec![], u32::MAX, u32::MAX).into(),
+        ])
+        .build()
+        .execute_with(|| {
+            let tank_funds = 100 * UNIT;
+            let max_tip = 1 * UNIT;
+            let para_id_with_tip = 1003u32;
+            let para_id_without_tip = 1001u32;
+
+            // Send funds to tank to both paras
+            assert_ok!(ServicesPayment::purchase_credits(
+                origin_of(ALICE.into()),
+                para_id_with_tip.into(),
+                tank_funds,
+            ));
+            assert_ok!(ServicesPayment::purchase_credits(
+                origin_of(ALICE.into()),
+                para_id_without_tip.into(),
+                tank_funds,
+            ));
+
+            // Only set tip for 1003
+            assert_ok!(ServicesPayment::set_max_tip(
+                root_origin(),
+                para_id_with_tip.into(),
+                max_tip,
+            ));
+
+            run_to_session(2);
```

Comment:
In this and the following test, for some reason, if it runs to the 1st session it seems that containers are not getting prioritized (but it does on the prior tests). But running to the 2nd session withdraws the tips corresponding to two sessions. It might have something to do for having the two additional collators (eve, ferdie)

Reply:
This is related to how assignment is calculated one session in advance. Changes to the list of para ids are only visible after 2 sessions. This explains why we only see the chain 1003 being assigned in the current assignment in session 2: when session 1 starts, the assignment for session 2 will be calculated. That assignment includes the changes to `max_tip`, so chain 1003 will be given priority. But that assignment will be applied in session 2.

The tip is being deducted twice: in session 1 (for the assignment in session 2), and in session 2 (for the assignment in session 3). In the previous tests the fee is only deduced once because we only `run_to_session(1)`, not 2 like here. You can verify this by adding this to the test:

```rust
run_to_session(1);
assert_eq!(
    Balances::usable_balance(ServicesPayment::parachain_tank(para_id_with_tip.into())),
    tank_funds - max_tip,
);
```